specutils/analysis/correlation.py

```diff
@@ -0,0 +1,119 @@
+
+
+def _normalize(observed_spectrum, template_spectrum):
+    """
+    Calculate a scale factor to be applied to the template spectrum so the
+    total flux in both spectra will be the same.
+
+    Parameters
+    ----------
+    observed_spectrum : :class:`~specutils.Spectrum1D`
+        The observed spectrum.
+    template_spectrum : :class:`~specutils.Spectrum1D`
+        The template spectrum, which needs to be normalized in order to be
+        compared with the observed spectrum.
+
+    Returns
+    -------
+    `float`
+        A float which will normalize the template spectrum's flux so that it
+        can be compared to the observed spectrum.
+    """
+    num = np.nansum((observed_spectrum.flux*template_spectrum.flux)/
+                 (observed_spectrum.uncertainty.array**2))
+    denom = np.nansum((template_spectrum.flux/
+                    observed_spectrum.uncertainty.array)**2)
+
+    return num/denom
+
+
+def template_correlate(observed_spectrum, template_spectrum):
+    """
+    Compute cross-correlation of the observed and template spectra.
+    This assumes that both are resampled into a common wavelength
+    scale before being submitted to this function.
+
+    Parameters
+    ----------
+    observed_spectrum : :class:`~specutils.Spectrum1D`
+        The observed spectrum.
+    template_spectrum : :class:`~specutils.Spectrum1D`
+        The template spectrum, which will be correlated with
+        the observed spectrum.
+
+    Returns
+    -------
+    (`~astropy.units.Quantity`, `~astropy.units.Quantity`)
+        Arrays with correlation values and lags in km/s
+    """
+    # Normalize template
+    normalization = _normalize(observed_spectrum, template_spectrum)
+
+    # Not sure if we need to actually normalize the template. Depending
+    # on the specific data uncertainty, the normalization factor
+    # may turn out negative. That causes a flip of the correlation function,
+    # in which the maximum (correlation peak) is no longer meaningful.
+    if normalization < 0.:
+        normalization = 1.
+
+    # Resample both spectrum and template to log wavelength. This is assuming
+    # that both spectrum and template share a common spectral axis.
+    resampler = FluxConservingResampler()
+    observed_log_spectrum = _log_resampler(observed_spectrum, resampler)
+    template_log_spectrum = _log_resampler(template_spectrum, resampler)
+
+    # Correlate
+    corr = np.correlate(observed_log_spectrum.flux.value,
+                        (template_log_spectrum.flux.value * normalization),
+                        mode='full')
+
+    # Compute lag in km/s
+    # wave_l is the wavelength array equally spaced in log space.
+    wave_l = observed_log_spectrum.spectral_axis.value
+    delta_log_wave = np.log10(wave_l[1]) - np.log10(wave_l[0])
+    deltas = (np.array(range(len(corr))) - len(corr)/2 + 0.5) * delta_log_wave
+    lags = (np.power(10., deltas) - 1. ) * const.c.to('km/s')
```

Comment:
Maybe the lags should be in "redshift" units instead of km/s?

Or alternatively, perhaps a keyword argument of `lag_units` that defaults to km/s, but can alternatively be a velocity unit (in which case you'd just replace this line with ``...to(lag_units)`` or the `dimensionless_unscaled` unit, which is then interpreted as redshift?  I.e. this line but without multiplying by `c`?

Reply:
This sounds easy to implement as well. The reason I have the lags done in that way is just because the IRAF xcorr task does it in that way. I used xcorr's code to guide me into implementing the steps in the notebook and function.